### PR TITLE
Speedup voxel grid build

### PR DESCRIPTION
**Pull request by anhnv3991**
- Use multithread to build the voxel grid of multigrid NDT omp
- Execution time of align in a test on shiojiri map
speedup_voxel_grid_build
- Execution time of updating NDT in the same test on shiojiri map
- 
speedup_update_ndt

### CODE REVIEW

**Comment by SakodaShintaro** [comment]
@anhnv3991 (cc. @YamatoAndo)  
Thank you for fixing the conflict.

There seems to be a mix of factors in this pull request.
I would like to proceed with the pull request in 4 steps.

(1) Refactor radiusSearch -> https://github.com/tier4/ndt_omp/pull/49 (Merged)
(2) Minor modifications -> https://github.com/tier4/ndt_omp/pull/50 (Merged)
(3) Change LeafID structure -> https://github.com/tier4/ndt_omp/pull/48 (Merged)
(4) Multithreading voxel grid building -> This PR

It seems that (3) contribute to speeding up align().
This pull request will help speed up Voxel construction.

**Comment by anhnv3991** [comment]
@SakodaShintaro I fixed all conflicts. Can you please review again?

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h [hunk 1]
```diff
@@ -142,20 +152,69 @@ class MultiVoxelGridCovariance : public pcl::VoxelGrid<PointT> {
...
+    }
+
+    Eigen::Vector3d &getEvals() {
+      return (evals_);
+    }
+
+    /** \brief Get the number of points contained by this voxel.
+     * \return number of points
+     */
+    int getPointCount() const {
```
Comment (SakodaShintaro) [inline include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:206]: Regarding `getcov()` to `getPointCount()`, these do not seem to be used, but are they necessary?

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h [hunk 2]
```diff
@@ -285,4 +411,4 @@ class MultiVoxelGridCovariance : public pcl::VoxelGrid<PointT> {
 };
 }  // namespace pclomp
 
-#endif  // #ifndef PCL_MULTI_VOXEL_GRID_COVARIANCE_H_
+#endif  // #ifndef PCL_MULTI_VOXEL_GRID_COVARIANCE_H_
```
Comment (SakodaShintaro) [inline include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: Could you please add a line break?

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp [hunk 3]
```diff
@@ -386,4 +410,4 @@ void pclomp::MultiVoxelGridCovariance<PointT>::computeLeafParams(const Eigen::Ve
 
 #define PCL_INSTANTIATE_VoxelGridCovariance(T) template class PCL_EXPORTS pcl::VoxelGridCovariance<T>;
 
-#endif  // PCL_MULTI_VOXEL_GRID_COVARIANCE_IMPL_H_
+#endif  // PCL_MULTI_VOXEL_GRID_COVARIANCE_IMPL_H_
```
Comment (SakodaShintaro) [inline include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp]: Could you please add a line break?

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp [hunk 4]
```diff
@@ -124,25 +136,36 @@ void MultiVoxelGridCovariance<PointT>::setInputCloudAndFilter(const PointCloudCo
...
   }
 
-  const auto [sid, iid] = *(sid_to_iid_.find(grid_id));
-
   // Set the pointer corresponding to the specified grid to null
-  grid_list_[iid].reset();
+  grid_list_[iid->second].reset();
   // Remove the specified grid from the conversion map
-  sid_to_iid_.erase(sid);
+  sid_to_iid_.erase(iid);
```
Comment (SakodaShintaro) [inline include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp:161]: I think it looked better before the change.

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h [hunk 5]
```diff
@@ -252,7 +314,65 @@ class MultiVoxelGridCovariance : public pcl::VoxelGrid<PointT> {
...
 protected:
+  // Return the index of an idle thread, which is not running any
+  // job, or has already finished its job and waiting for a join.
+  inline int get_idle_tid() {
+    int tid = (last_check_tid_ == thread_num_ - 1) ? 0 : last_check_tid_ + 1;
+    std::chrono::microseconds span(50);
+
+    // Loop until an idle thread is found
+    while(true) {
+      // gettimeofday(&start, NULL);
```
Comment (SakodaShintaro) [inline include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: Is this line unnecessary?

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h [hunk 6]
```diff
@@ -252,7 +314,65 @@ class MultiVoxelGridCovariance : public pcl::VoxelGrid<PointT> {
   // Return the string indices of currently loaded map pieces
   std::vector<std::string> getCurrentMapIDs() const;
 
+  void setThreadNum(int thread_num) {
+    if(thread_num <= 0) {
+      thread_num_ = 1;
+    }
+
+    thread_num_ = thread_num;
+    thread_futs_.resize(thread_num_);
```
Comment (SakodaShintaro) [inline include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:317]: Just to confirm, is sync() not necessary before resizing?
I think that there might be a possibility of calling setThreadNum immediately after calling setInputCloudAndFilter.

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h [hunk 7]
```diff
@@ -61,6 +61,16 @@
...
 #include <Eigen/Cholesky>
+
+#include <future>
+#include <ctime>
+#include <fstream>
+#include <sstream>
+#include <sys/time.h>
+#include <pcl/io/pcd_io.h>
+#include <queue>
+#include <unistd.h>
```
Comment (SakodaShintaro) [inline include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: If there are any unnecessary items in the include list, please reduce them.

**Review by SakodaShintaro — CHANGES_REQUESTED** [review]
Thank you for fixing conflicts.
Please make some corrections.

It might be better to put all the definition code such as get_idle_tid() or sync() in _impl.h instead of .h, but I'll leave it up to you whether do it.

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h (hunk 1 above)
Comment (anhnv3991) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:206]: @SakodaShintaro In general we do not know which getCov() may be used (we may want to modify cov in some cases), so I want to keep both const and non-const getCov().
As for getPointCloud() it is just for completeness of the struct, though.

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h (hunk 2 above)
Comment (anhnv3991) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: @SakodaShintaro Done!

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp (hunk 4 above)
Comment (anhnv3991) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp:161]: @SakodaShintaro If we use this 
```  
if(sid_to_iid_.count(grid_id) == 0) {
    return;
 }

 const auto [sid, iid] = *(sid_to_iid_.find(grid_id));
```
we will have to search for grid_id twice. One by count() and one by find().
But if we use this 
```
  auto iid = sid_to_iid_.find(grid_id);

  if(iid == sid_to_iid_.end()) {
    return;
  }
```
only one search is performed. While the cost difference is negligible (the map is too small), I prefer not to perform the same action multiple times if not necessary.

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp (hunk 3 above)
Comment (anhnv3991) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp]: @SakodaShintaro Done!

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h (hunk 5 above)
Comment (anhnv3991) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: @SakodaShintaro I deleted it.

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h (hunk 6 above)
Comment (anhnv3991) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:317]: @SakodaShintaro Thanks. I added a sync in the begginning of setThreadNum.

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h (hunk 7 above)
Comment (anhnv3991) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: @SakodaShintaro Those are for debugging. I deleted them.

**Comment by anhnv3991** [comment]
> Thank you for fixing conflicts. Please make some corrections.
> 
> It might be better to put all the definition code such as get_idle_tid() or sync() in _impl.h instead of .h, but I'll leave it up to you whether do it.

@SakodaShintaro Those functions are short and called many times so I prefer to keep them inline to minimize the cost of calling.

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h [hunk 8]
```diff
@@ -252,7 +306,66 @@ class MultiVoxelGridCovariance : public pcl::VoxelGrid<PointT> {
...
+      }
+
+      // If the current thread has not finished its job, check the next thread
+      tid = (tid == thread_num_ - 1) ? 0 : tid + 1;
+    }
+  }
+
+  // Wait for all running threads to finish
+  inline void sync() {
+    for(int i = 0; i < thread_futs_.size(); ++i) {
```
Comment (SakodaShintaro) [inline include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: Sorry, additional comment.
autoware.universe strictly checks warnings, so when I built `ndt_scan_matcher`, the following error occurred.

```
+ colcon build --symlink-install --cmake-args -DCMAKE_BUILD_TYPE=Release --packages-up-to ndt_scan_matcher
Starting >>> ndt_scan_matcher
[Processing: ndt_scan_matcher]                               
--- stderr: ndt_scan_matcher                                     
In file included from /home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:62,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/hyper_parameters.hpp:20,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/map_update_module.hpp:19,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/map_update_module.cpp:15:
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h: In instantiation of 'void pclomp::MultiVoxelGridCovariance<PointT>::sync() [with PointT = pcl::PointXYZ]':
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:310:5:   required from 'void pclomp::MultiVoxelGridCovariance<PointT>::setThreadNum(int) [with PointT = pcl::PointXYZ]'
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:321:31:   required from 'void pclomp::MultiGridNormalDistributionsTransform<PointSource, PointTarget>::setParams(const pclomp::NdtParams&) [with PointSource = pcl::PointXYZ; PointTarget = pcl::PointXYZ]'
/home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/map_update_module.cpp:79:24:   required from here
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:354:22: error: comparison of integer expressions of different signedness: 'int' and 'std::vector<std::future<bool> >::size_type' {aka 'long unsigned int'} [-Werror=sign-compare]
  354 |     for(int i = 0; i < thread_futs_.size(); ++i) {
      |                    ~~^~~~~~~~~~~~~~~~~~~~~
In file included from /home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:62,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/hyper_parameters.hpp:20,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/map_update_module.hpp:19,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/map_update_module.cpp:15:
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h: In instantiation of 'void pclomp::MultiVoxelGridCovariance<PointT>::sync() [with PointT = pcl::PointXYZ]':
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:310:5:   required from 'void pclomp::MultiVoxelGridCovariance<PointT>::setThreadNum(int) [with PointT = pcl::PointXYZ]'
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:321:31:   required from 'void pclomp::MultiGridNormalDistributionsTransform<PointSource, PointTarget>::setParams(const pclomp::NdtParams&) [with PointSource = pcl::PointXYZ; PointTarget = pcl::PointXYZ]'
/home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/map_update_module.cpp:79:24:   required from here
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:354:22: error: comparison of integer expressions of different signedness: 'int' and 'std::vector<std::future<bool> >::size_type' {aka 'long unsigned int'} [-Werror=sign-compare]
  354 |     for(int i = 0; i < thread_futs_.size(); ++i) {
      |                    ~~^~~~~~~~~~~~~~~~~~~~~
In file included from /home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:62,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/hyper_parameters.hpp:20,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/ndt_scan_matcher_core.hpp:21,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:15:
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h: In instantiation of 'void pclomp::MultiVoxelGridCovariance<PointT>::sync() [with PointT = pcl::PointXYZ]':
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:310:5:   required from 'void pclomp::MultiVoxelGridCovariance<PointT>::setThreadNum(int) [with PointT = pcl::PointXYZ]'
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:321:31:   required from 'void pclomp::MultiGridNormalDistributionsTransform<PointSource, PointTarget>::setParams(const pclomp::NdtParams&) [with PointSource = pcl::PointXYZ; PointTarget = pcl::PointXYZ]'
/home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:178:22:   required from here
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:354:22: error: comparison of integer expressions of different signedness: 'int' and 'std::vector<std::future<bool> >::size_type' {aka 'long unsigned int'} [-Werror=sign-compare]
  354 |     for(int i = 0; i < thread_futs_.size(); ++i) {
      |                    ~~^~~~~~~~~~~~~~~~~~~~~
In file included from /home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:62,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/hyper_parameters.hpp:20,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/ndt_scan_matcher_core.hpp:21,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:15:
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h: In instantiation of 'void pclomp::MultiVoxelGridCovariance<PointT>::sync() [with PointT = pcl::PointXYZ]':
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:310:5:   required from 'void pclomp::MultiVoxelGridCovariance<PointT>::setThreadNum(int) [with PointT = pcl::PointXYZ]'
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:321:31:   required from 'void pclomp::MultiGridNormalDistributionsTransform<PointSource, PointTarget>::setParams(const pclomp::NdtParams&) [with PointSource = pcl::PointXYZ; PointTarget = pcl::PointXYZ]'
/home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:178:22:   required from here
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:354:22: error: comparison of integer expressions of different signedness: 'int' and 'std::vector<std::future<bool> >::size_type' {aka 'long unsigned int'} [-Werror=sign-compare]
  354 |     for(int i = 0; i < thread_futs_.size(); ++i) {
      |                    ~~^~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
gmake[2]: *** [CMakeFiles/once_initialize_at_out_of_map_then_initialize_correctly.dir/build.make:118: CMakeFiles/once_initialize_at_out_of_map_then_initialize_correctly.dir/src/map_update_module.cpp.o] エラー 1
gmake[2]: *** 未完了のジョブを待っています....
cc1plus: all warnings being treated as errors
gmake[2]: *** [CMakeFiles/standard_sequence_for_initial_pose_estimation.dir/build.make:118: CMakeFiles/standard_sequence_for_initial_pose_estimation.dir/src/map_update_module.cpp.o] エラー 1
gmake[2]: *** 未完了のジョブを待っています....
cc1plus: all warnings being treated as errors
gmake[2]: *** [CMakeFiles/once_initialize_at_out_of_map_then_initialize_correctly.dir/build.make:104: CMakeFiles/once_initialize_at_out_of_map_then_initialize_correctly.dir/src/ndt_scan_matcher_core.cpp.o] エラー 1
gmake[1]: *** [CMakeFiles/Makefile2:213: CMakeFiles/once_initialize_at_out_of_map_then_initialize_correctly.dir/all] エラー 2
gmake[1]: *** 未完了のジョブを待っています....
cc1plus: all warnings being treated as errors
gmake[2]: *** [CMakeFiles/standard_sequence_for_initial_pose_estimation.dir/build.make:104: CMakeFiles/standard_sequence_for_initial_pose_estimation.dir/src/ndt_scan_matcher_core.cpp.o] エラー 1
gmake[1]: *** [CMakeFiles/Makefile2:186: CMakeFiles/standard_sequence_for_initial_pose_estimation.dir/all] エラー 2
gmake: *** [Makefile:146: all] エラー 2
---
Failed   <<< ndt_scan_matcher [42.0s, exited with code 2]

Summary: 34 packages finished [1min 2s]
  1 package failed: ndt_scan_matcher
  1 package had stderr output: ndt_scan_matcher
```

Comment (anhnv3991) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: @SakodaShintaro I fixed this one but do you know why this error did not show up when I build the code on my local PC?

Comment (SakodaShintaro) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: Hmm, the only thing I can think of is that the build cache remained. Thank you for the fixing.

**Comment by SakodaShintaro** [comment]
Thank you for correcting the code.
There is no problem with the code, so I will test it carefully.
I will check the following items
- [x] unit test of ndt_scan_matcher in autoware.universe
- [x] logging_simulator (sample-rosbag)
- [x] logging_simulator (AWSIM-rosbag) single map with comparing to GT Pose
- [x] logging_simulator (AWSIM-rosbag) divided map with comparing to GT Pose
- [x] driving_log_replayer
- [x] AWSIM (1 hour, 20 loops OK)

Please wait a moment.

**Comment by anhnv3991** [comment]
@SakodaShintaro Thanks a lot!

**Review by SakodaShintaro — APPROVED** [review]
Looks Good To Me
Marvelous work!

**Pull request merged**